program/src/processor/swap.rs

```diff
@@ -0,0 +1,364 @@
+    pub user_quote_account: &'a T,
+    #[cons(writable, signer)]
+    pub user_owner: &'a T,
+    pub discount_token_account: Option<&'a T>,
+}
+
+impl<'a, 'b: 'a> Accounts<'a, AccountInfo<'b>> {
+    pub fn parse(
+        program_id: &Pubkey,
+        accounts: &'a [AccountInfo<'b>],
+    ) -> Result<Self, ProgramError> {
+        let accounts_iter = &mut accounts.iter();
+        let a = Self {
+            spl_token_program: next_account_info(accounts_iter)?,
+            system_program: next_account_info(accounts_iter)?,
+            market: next_account_info(accounts_iter)?,
+            orderbook: next_account_info(accounts_iter)?,
+            event_queue: next_account_info(accounts_iter)?,
+            bids: next_account_info(accounts_iter)?,
+            asks: next_account_info(accounts_iter)?,
+            base_vault: next_account_info(accounts_iter)?,
+            quote_vault: next_account_info(accounts_iter)?,
+            market_signer: next_account_info(accounts_iter)?,
+            user_base_account: next_account_info(accounts_iter)?,
+            user_quote_account: next_account_info(accounts_iter)?,
+            user_owner: next_account_info(accounts_iter)?,
+            discount_token_account: next_account_info(accounts_iter).ok(),
+        };
+        check_signer(a.user_owner).map_err(|e| {
+            msg!("The user account owner should be a signer for this transaction!");
+            e
+        })?;
+        check_account_key(
+            a.spl_token_program,
+            &spl_token::ID.to_bytes(),
+            DexError::InvalidSplTokenProgram,
+        )?;
+        check_account_key(
+            a.system_program,
+            &system_program::ID.to_bytes(),
+            DexError::InvalidSystemProgramAccount,
+        )?;
+        if let Some(discount_account) = a.discount_token_account {
+            check_account_owner(
+                discount_account,
+                a.user_owner.key,
+                DexError::InvalidStateAccountOwner,
+            )?;
+        }
+        check_account_owner(a.market, program_id, DexError::InvalidStateAccountOwner)?;
+
+        Ok(a)
+    }
+}
+
+pub(crate) fn process(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    instruction_data: &[u8],
+) -> ProgramResult {
+    let accounts = Accounts::parse(program_id, accounts)?;
+
+    let Params {
+        side,
+        base_qty,
+        mut quote_qty,
+        self_trade_behavior,
+        match_limit,
+        _padding: _,
+    } = try_from_bytes(instruction_data).map_err(|_| ProgramError::InvalidInstructionData)?;
+    let market_state = DexState::get(accounts.market)?;
+
+    // Check the order size
+    if base_qty < &market_state.min_base_order_size {
+        msg!("The base order size is too small.");
+        return Err(ProgramError::InvalidArgument);
+    }
+
+    check_accounts(program_id, &market_state, &accounts).unwrap();
+    let callback_info = CallBackInfo {
+        user_account: Pubkey::default(),
```

Comment:
maybe can change callbackinfo to use Option<Pubkey> instead, but that felt messy (had to update a bunch of places, change CALLBACK_INFO_LEN, etc.)